### ISSUE TITLE
Attacher doesn't work with symbolized keys

### ISSUE THREAD

**Issue by Morozzzko** [shrinerb/shrine#337]
I have a simple Hanami app with a tiny uploader (`class ImageUploader < Shrine`).

I can't use `include ImageUploader::Attachment[:image]` in my entities (`Hanami::Model`) since **all entities are frozen**, so I decided to write a simple method called `image_attacher`.

```
class Drawing < Hanami::Entity
  def image_attacher
    ImageUploader::Attacher.new(self, 'image')
  end
end
```

When I try to call the method, it says  ```
Shrine::Error: {:id=>"a5185fa0d61ade5fac7ec75804db35a3.png", :storage=>"store", :metadata=>{:size=>183774, :filename=>"RackMultipart20181217-63192-4bleet.png", :mime_type=>"image/png"}} isn't valid uploaded file data```

So I went inside the sources and realized that Shrine expects the hash keys to be stringified:

https://github.com/shrinerb/shrine/blob/950a12056b63706c1bdd949bb7c647b57cb4ac0b/lib/shrine/uploaded_file.rb#L31

Feels like an inconvenience that should be either fixed or documented. Quick search didn't show anything

**Comment by janko-m** [comment]
Yeah, Shrine expects that you'll be persisting your data in a string/JSON-like DB column, which doesn't support Ruby symbols. I agree that it should be documented. I will also consider adding support for symbols in the near future.

**Comment by Morozzzko** [comment]
How do you think that support should work? 

I see multiple options:

* Forcefully coerce data to symbolized/stringified hash
* Use something like `HashWithIndifferentAccess` 
* Use a function that tries to fetch using both symbol and string keys
* Same as the function, except it's a `refinement`

Personally, I would stick to the third option. I might send a PR within next two weeks if we decide on the implementation

**Comment by janko-m** [comment]
After thinking about it more, I would like to avoid supporting symbol keys for the time being, as I feel like the added complexity is not worth it. In addition to modifying `#id`, `#storage_key`, and `#metadata`, some plugins like `refresh_metadata` would also need to be modified to handle symbol keys as well. I feel like I'd be bound to forget handling symbol keys when implementing some feature, which would then cause this feature not working properly with symbol keys.

**Comment by janko-m** [comment]
Yeah, let's move it to the Google group 👍

**Comment by janko** [comment]
I'm currently refactoring `Shrine::Attacher`, and would like for `Shrine::UploadedFile` data hash to support symbol keys in the near future (string keys couple to JSON anyway). So, this is a completely valid request, sorry that I closed it. Now that Shrine dropped MRI 2.1 support, we can rely on Symbol GC.

**Comment by janko** [comment]
This has been addressed in https://github.com/shrinerb/shrine/commit/b3fba16bb1d072384beac4f179cdc01a6a96b161. `Shrine.uploaded_file` method now accepts hashes with symbol keys, in which case it will deep-stringify the keys. This means loading attached files from column attribute with symbolized keys will automatically work.

This will be released in Shrine 3.0.
